### PR TITLE
refactor!: update operations to use delta scan

### PR DESCRIPTION
**Pull request by Blajda**
# Description
Recently implemented operations did not use `DeltaScan` it had some gaps. These gaps would make it harder switch towards logical plans which is required for merge.

Gaps:
- It was not possible to include file lineage in the result
- The subset of files to be scanned is known ahead of time. Users had to reconstruct a parquet scan based on those files

The PR introduces a `DeltaScanBuilder` that allow users to specify which files to use when constructing the scan, if the scan should be enhanced to include additional metadata columns, and allows a projection to be specified. It also retains previous functionality of pruning based on the provided filter when files to scan are not provided.

`DeltaScanConfig` is also introduced which allows users to deterministic obtain the names of any added metadata columns or allows them to specify the name if required.

The public interface for `find_files` has changed but functionality remains the same.

A new table provider was introduced which accepts an `DeltaScanConfig`. This is required for future merge enhancements so unmodified files can be pruned pruned prior to writes.

### CODE REVIEW

**Comment by Blajda** [comment]
Can be merged after #1705

**Review by wjones127 — DISMISSED** [review]

**Comment by Blajda** [comment]
@wjones127 This is ready to be merged

File: rust/src/delta_datafusion/mod.rs [hunk 1]
```diff
@@ -362,66 +364,299 @@ pub(crate) fn register_store(store: ObjectStoreRef, env: Arc<RuntimeEnv>) {
...
+            arrow_schema::DataType::Utf8,
+            true,
+        )));
+    }
+
+    Ok(Arc::new(ArrowSchema::new(fields)))
+}
+
+#[derive(Debug, Clone, Default)]
+/// Used to specify it additonal metadat columns are exposed to the user
```
Comment (roeap) [inline rust/src/delta_datafusion/mod.rs:389]:
```suggestion
/// Used to specify if additonal metadat columns are exposed to the user
```

File: rust/src/delta_datafusion/mod.rs [hunk 2]
```diff
@@ -362,66 +364,299 @@ pub(crate) fn register_store(store: ObjectStoreRef, env: Arc<RuntimeEnv>) {
...
+}
+
+impl DeltaScanConfigBuilder {
+    /// Construct a new instance of `DeltaScanConfigBuilder`
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Indicate that a column containing a records file path is included.
+    /// Column name is randomly generated and can be determined once this Config is built
```
Comment (roeap) [inline rust/src/delta_datafusion/mod.rs:407]:
```suggestion
    /// Column name is generated and can be determined once this Config is built
```

**Review by roeap — COMMENTED** [review]
Looking great so far! Running out of time for now, but will finish asap!

**Review by roeap — APPROVED** [review]
LGTM! operations are really coming along :).

**Pull request merged**